### PR TITLE
Updates for newer Ginkgo versions and improvement to Ginkgo OpenMP use

### LINKED ISSUE #3909: Resolve Ginkgo warnings and fix Ginkgo backend selection
1. When testing MFEM 4.6 with Ginkgo 1.6.0, I noticed a lot of warnings about deprecation:
    ```console
    In file included from linalg/linalg.hpp:49,
                     from linalg/sparsemat.cpp:14:
    linalg/ginkgo.hpp: In member function 'void mfem::Ginkgo::ResidualLogger<ValueType>::on_iteration_complete(const gko::LinOp*, const size_type&, const gko::LinOp*, const gko::LinOp*, const gko::LinOp*, const gko::LinOp*) const':
    linalg/ginkgo.hpp:339:42: warning: 'typename std::enable_if<(! gko::detail::have_ownership_impl<typename std::decay<_Tp>::type>::value), typename std::remove_reference<decltype (* declval<typename std::decay<_Tp>::type>())>::type*>::type gko::lend(const Pointer&) [with Pointer = const gko::LinOp*; typename std::enable_if<(! gko::detail::have_ownership_impl<typename std::decay<_Tp>::type>::value), typename std::remove_reference<decltype (* declval<typename std::decay<_Tp>::type>())>::type*>::type = const gko::LinOp*]' is deprecated: no longer necessary, just pass the object without lend [-Wdeprecated-declarations]
              matrix->apply(gko::lend(solution), gko::lend(res));
                                              ^
    ```

2. Fix the Ginkgo backend selection when MFEM uses "cpu" deveices -- Ginkgo should use its reference (single thread) backend instead of OpenMP. The OpenMP Ginkgo backend should be used when MFEM uses "omp" device. See the implementation here: https://github.com/mfem/mfem/blob/c0867ca009e9b5fafb5585318648f430b1d6eef0/linalg/ginkgo.cpp#L82
and, for some context, see the following two comments:
    * https://github.com/xsdk-project/xsdk-examples/pull/37#issuecomment-1532444883, and
    * https://github.com/xsdk-project/xsdk-examples/pull/37#issuecomment-1533563468;

### PR DESCRIPTION
**Pull request by nbeams**
This PR is a proposal to address #3909. 

There are changes to the use of `gko::lend` (deprecated) and the Ginkgo loggers, with compiler guards based on the version of Ginkgo being used. Hopefully all the deprecation warnings should be gone now. I built locally with Ginkgo 1.5 and 1.6 to test.

For Ginkgo's use of OpenMP, ~~a small tester has been added in `config` to determine whether or not Ginkgo was built with OpenMP. I have tried to closely follow the way other testers are used (e.g., `get_hypre_version`) for both the CMake and Makefile builds, but please let me know if I've done something unacceptable to the build setup.~~ the build status of Ginkgo is now checked at runtime.

For the default selection of Ginkgo executors, the proposed flow goes like this: 
| MFEM device type | Ginkgo Executor | Ginkgo Host Executor
|:-----------------|:-----------------:|:----------------:|
| CPU | OpenMP if Ginkgo has it *and* the MFEM device uses OpenMP; otherwise, Reference | N/A |
| CUDA | Cuda | OpenMP if Ginkgo has OpenMP, otherwise, Reference | 
| HIP   | Hip | OpenMP if Ginkgo has OpenMP, otherwise, Reference | 

I also added new constructors to `GinkgoExecutor`, allowing the user to explicitly set the host Executor instead of going with the default, e.g. if Ginkgo has OpenMP, but the user wants a `ReferenceExecutor` to pair with a `CudaExecutor`.

Closes #3909.

 | PR | Author | Editor | Reviewers | Assignment | Approval | Merge|
 | --- | --- | --- | --- | --- | --- | --- |
| [#3934](https://github.com/mfem/mfem/pull/3934) | @nbeams | @v-dobrev | @v-dobrev + @sebastiangrimberg | 10/18/23 | 10/31/23 | 11/6/23 | |

<details>
<summary>PR Checklist</summary>
    
- [ ] Code builds.
- [ ] Code passes `make style`.
- [ ] Update `CHANGELOG`:
    - [ ] Is this a new feature users need to be aware of? New or updated example or miniapp?
    - [ ] Does it make sense to create a new section in the `CHANGELOG` to group with other related features?
- [ ] Update `INSTALL`:
    - [ ] Had a new optional library been added? If so, what range of versions of this library are required? (*Make sure the external library is compatible with our BSD license, e.g. it is not licensed under GPL!*)
    - [ ] Have the version ranges for any required or optional libraries changed?
    - [ ] Does `make` or `cmake` have a new target?
    - [ ] Did the requirements or the installation process change? *(rare)*
- [ ] Update continuous integration server configurations if necessary (e.g. with new version requirements for each of MFEM's dependencies)
    - [ ] `.github`
    - [ ] `.appveyor.yml`
- [ ] Update `.gitignore`:
    - [ ] Check if `make distclean; git status` shows any files that were generated from the source by the project (not an IDE) but we don't want to track in the repository.
    - [ ] Add new patterns (just for the new files above) and re-run the above test.
- [ ] New examples:
    - [ ] All sample runs at the top of the example source file work.
    - [ ] Update `examples/makefile`:
      - [ ] Add the example code to the appropriate `SEQ_EXAMPLES` and `PAR_EXAMPLES` variables.
      - [ ] Add any files generated by it to the `clean` target.
      - [ ] Add the example binary and any files generated by it to the top-level `.gitignore` file.
    - [ ] Update `examples/CMakeLists.txt`:
      - [ ] Add the example code to the `ALL_EXE_SRCS` variable.
      - [ ] Make sure `THIS_TEST_OPTIONS` is set correctly for the new example.
   - [ ] List the new example in `doc/CodeDocumentation.dox`.
   - [ ] If new examples directory (e.g.`examples/pumi`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
      - [ ] Update or add example-specific documentation, see e.g. the `src/examples.md`.
      - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
      - [ ] In `examples.md`, list the example under the appropriate categories, add new categories if necessary.
      - [ ] Add a short description of the example in the "Extensive Examples" section of `features.md`.
- [ ] New miniapps:
   - [ ] All sample runs at the top of the miniapp source file work.
   - [ ] Update top-level `makefile` and `makefile` in corresponding miniapp directory.
   - [ ] Add the miniapp binary and any files generated by it to the top-level `.gitignore` file.
   - [ ] Update CMake build system:
      - [ ] Update the `CMakeLists.txt` file in the `miniapps` directory, if the new miniapp is in a new directory.
      - [ ] Add/update the `CMakeLists.txt` file in the new miniapp directory.
      - [ ] Consider adding a new test for the new miniapp.
   - [ ] List the new miniapp in `doc/CodeDocumentation.dox`
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), add it to `MINIAPP_SUBDIRS` in the `makefile`.
   - [ ] If new miniapps directory (e.g.`miniapps/nurbs`), list it in `doc/CodeDocumentation.conf.in`
   - [ ] Companion pull request for documentation in [mfem/web](https://github.com/mfem/web) repo:
     - [ ] Update or add miniapp-specific documentation, see e.g. the `src/meshing.md` and `src/electromagnetics.md` files.
     - [ ] Add the description, labels and screenshots in `src/examples.md` and `src/img`.
     - [ ] The miniapps go at the end of the page, and are usually listed only under a specific "Application (PDE)" category.
     - [ ] Add a short description of the miniapp in the "Extensive Examples" section of `features.md`.
- [ ] New capability:
   - [ ] All new public, protected, and private classes, methods, data members, and functions have full Doxygen-style documentation in source comments. Documentation should include descriptions of member data, function arguments and return values, template parameters, and prerequisites for calling new functions.
   - [ ] Pointer arguments and return values must specify whether ownership is being transferred or lent with the call.
   - [ ] Any new functions should include descriptions of their intended use e.g. for internal use only, user-facing, etc., along with references to example code whenever possible/appropriate.
   - [ ] Consider adding new sample runs in existing examples to highlight the new capability.
   - [ ] Consider saving cool simulation pictures with the new capability in the Confluence gallery (LLNL only) or submitting them, via pull request, to the gallery section of the `mfem/web` repo.
   - [ ] If this is a major new feature, consider mentioning it in the short summary inside `README` *(rare)*.
   - [ ] List major new classes in `doc/CodeDocumentation.dox` *(rare)*.
- [ ] Update this checklist, if the new pull request affects it.
- [ ] Run `make unittest` to make sure all unit tests pass.
- [ ] Run the tests in `tests/scripts`.
- [ ] (LLNL only) After merging:
   - [ ] Update internal tests to include the new features.
</details>

### CODE REVIEW

File: config/get_ginkgo_omp_config.cpp
Comment (v-dobrev) [inline config/get_ginkgo_omp_config.cpp:14]: For consistency, let's use `<cstring>` instead.

**Comment by v-dobrev** [comment]
This PR is now [under review](https://github.com/mfem/mfem/blob/master/CONTRIBUTING.md#pull-requests) (see the table in the PR description). To help with the review process, please **do not force push** to the branch.

File: config/config.hpp.in
Comment (sebastiangrimberg) [inline config/config.hpp.in:107]: Maybe `MFEM_GINKO_USE_OPENMP` instead of `*_WITH_OMP` to be consistent with the definitions elsewhere? Would apply to `CMakeLists.txt` as well as the other GNU make related files.

File: config/get_ginkgo_omp_config.cpp
Comment (sebastiangrimberg) [inline config/get_ginkgo_omp_config.cpp:19]: Maybe I am missing something here, and not fully following the discussion in https://github.com/mfem/mfem/issues/3909. But, why do we need a test program and the compile time `MFEM_GINKGO_WITH_OMP` configuration variable if we can just use `gko::version_info::get` along with the MFEM `Device` configuration to determine whether to use `gko::ReferenceExecutor::create` vs. `gko::OmpExecutor::create` at runtime?

For example, with no compile-time configuration, I am thinking of the following type of thing in the `GinkgoExecutor` constructor:

```
gko::version_info gko_version = gko::version_info::get();
bool ginko_use_openmp = (strcmp(gko_version.omp_version.tag, "not compiled") == 0)
if (Device::Allows(Backend::OMP_MASK) && ginko_use_omp)
{
   executor = gko::OmpExecutor::create();
}
else
{
   executor = gko::ReferenceExecutor::create();
}
```

Similar logic could be applied for CUDA/HIP device support. Basically it seems like everything can be done at runtime here.

Comment (nbeams) [reply config/get_ginkgo_omp_config.cpp:19]: Mostly I just figured, if we could settle the issue at compile time, why not go ahead and do it? But I guess it doesn't really matter, as it's not critical for performance or anything. I can switch to a runtime check in the constructors if that would be preferred. It would certainly avoid adding complexity to the build system.

Comment (sebastiangrimberg) [reply config/get_ginkgo_omp_config.cpp:19]: I'm in favor of keeping the build system as simple as possible (only a matter of time before users come asking about errors encountered during `try_run`). This seems like a good fit for runtime configuration and is consistent with the rest of the MFEM approach of device configuration at run time, not build time.

**Comment by nbeams** [comment]
Oh no -- I'm not supposed to force-push to the branch during review, but I just realized I'm an idiot who put "*to* compile-time" instead of "*from* compile-time" in the commit message. I need to fix it before we merge this branch because it's a completely incorrect description of the commit as-is. Can I force-push just this once?  🤦‍♀️

File: linalg/ginkgo.hpp
Comment (sebastiangrimberg) [inline linalg/ginkgo.hpp:49]: My first reaction here is that we should avoid `#define` in headers which are installed by MFEM like this one since it will affect all downstream user code. We could move some of these function definitions and the macro into `ginkgo.cpp` to fix this. But on second thought maybe these are safe, and moving things into the `.cpp` will be more trouble than necessary. Maybe @v-dobrev do you have an opinion here?

For the template functions you may just need to explicitly instantiate them in the `.cpp` file since they aren't being defined in the same file where they are declared. For example:

```
template double Ginkgo::compute_norm(const gko::matrix::Dense<double> *);
```

File: linalg/ginkgo.cpp
Comment (sebastiangrimberg) [inline linalg/ginkgo.cpp:61]: `if (gko_omp_status)` (and elsewhere in this file)

File: linalg/ginkgo.cpp
Comment (sebastiangrimberg) [inline linalg/ginkgo.cpp:36]: `bool gko_omp_status = (strcmp(gko_version.omp_version.tag, "not compiled") != 0)`

File: linalg/ginkgo.cpp
Comment (sebastiangrimberg) [inline linalg/ginkgo.cpp:515]: Remove the added newline

File: linalg/ginkgo.cpp
Comment (sebastiangrimberg) [inline linalg/ginkgo.cpp:354]: Replace the deleted newline

File: linalg/ginkgo.cpp
Comment (sebastiangrimberg) [inline linalg/ginkgo.cpp:1064]: Remove the added newline

**Review by sebastiangrimberg — COMMENTED** [review]
Thanks @nbeams this looks good. Regarding your comment on force-pushing, I will defer to the editor @v-dobrev on how he wants to handle this. I personally think it would be fine to force push the last commit to revise the message but it's up to him.

Mostly all very minor comments, the only one with substance is wondering if we should try to move some function definitions out of `ginkgo.hpp` and into `ginkgo.cpp` so we can avoid defining preprocessor macros in the header file which then get consumed by everyone including `ginkgo.hpp`.

File: linalg/ginkgo.hpp
Comment (nbeams) [reply linalg/ginkgo.hpp:49]: I could also just put two separate lines of code with the compiler guards (one line with the `gko::lend`, one without). It's a bit messier to read, but it would avoid any `#define`s. Would that be better?

Comment (sebastiangrimberg) [reply linalg/ginkgo.hpp:49]: That seems fine. I was trying to decide if we could avoid defining `MFEM_GINKGO_VERSION` in the header too but this one is probably fine and is based on values from the Ginkgo headers rather than other compile time definitions. So I think your solution would be OK!

**Comment by nbeams** [comment]
@sebastiangrimberg thanks for the quick review, it will be simple to address your comments -- but I'd like to see about amending the previous commit message before I add another commit (@v-dobrev ?). Sorry again for the annoyance on that...

**Comment by v-dobrev** [comment]
@nbeams, in general, we prefer to avoid force-pushes during review (to make it easier for reviewers to track changes in the PR). However, if you just want to change the message in your last commit, force-pushing a correction will not be confusing, so it is fine to do it.

File: linalg/ginkgo.hpp
Comment (v-dobrev) [reply linalg/ginkgo.hpp:49]: I think having macros in MFEM headers is fine, however, I'd add a prefix `MFEM_` (`GKO_LEND` -> `MFEM_GKO_LEND`) since it is a macro defined by MFEM. Of course, removing the macro is also fine, if there is a preference for that approach -- I'm fine with either choice.

**Comment by nbeams** [comment]
I hope I addressed all the comments in [1ae3aad](https://github.com/mfem/mfem/pull/3934/commits/1ae3aad33695acc6b7df6eccfbb29f5d99826e31).

File: linalg/ginkgo.cpp
Comment (nbeams) [reply linalg/ginkgo.cpp]: I made this change, but I switched the variable name to `gko_with_omp_support` because I think it reads more clearly in the "short" form of `if (var)`.

File: linalg/ginkgo.cpp
Comment (nbeams) [reply linalg/ginkgo.cpp]: This was fixed in 1ae3aad

File: linalg/ginkgo.hpp
Comment (nbeams) [reply linalg/ginkgo.hpp]: Based on discussion here, I went ahead and removed the use of `GKO_LEND` but left `MFEM_GINKGO_VERSION` in ginkgo.hpp.

File: config/get_ginkgo_omp_config.cpp
Comment (nbeams) [reply config/get_ginkgo_omp_config.cpp]: Even though the tester has been removed, I used `<cstring>` in `ginkgo.hpp` where we now do the same check at runtime.

**Review by sebastiangrimberg — APPROVED** [review]

**Review by v-dobrev — APPROVED** [review]

**Comment by v-dobrev** [comment]
Merged in `next` for testing...

**Pull request merged**
